### PR TITLE
refactor: Make IOKey dataclass derived from BaseKey

### LINKED ISSUE #116: [OTHER] Rename shape method of TemplateBase
## Summary

Rename `shape` method name in TemplateBase.

## Context
- Convert `shape` method name to `get_shape` in TemplateBase.
- Update corresponding tests.

### PR DESCRIPTION
**Pull request by kberat-synnada**
## Description

Closes #116 and #118.

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: mithril/framework/common.py [hunk 1]
```diff
@@ -1001,7 +1001,7 @@ def abs(self):
     def len(self):
         return ExtendTemplate(connections=[self], model="len")
 
-    def shape(self):
+    def get_shape(self):
```
Comment (aturker-synnada) [inline mithril/framework/common.py:1004]: Are we sure about this API. I think it shape should add shape model, and get_shapes method should return attribute _shape

File: mithril/framework/common.py [hunk 2]
```diff
@@ -1100,49 +1100,41 @@ def __init__(
...
 
-        connections = connections or []
-        for item in connections:
-            conn: ConnectionData | str
-            conn = item.data if isinstance(item, Connection) else item
-            self._connections.add(conn)
-
-    def __hash__(self) -> int:
-        return hash(id(self))
+    def __eq__(self, other: object):
```
Comment (aturker-synnada) [inline mithril/framework/common.py:1133]: Why do we need to override __eq__ method only in IOKey?

Comment (kberat-synnada) [reply mithril/framework/common.py:1133]: Because dataclass overrides eq bt we want to return ExtendTemplate

**Review by norhan-synnada — APPROVED** [review]

**Review by aturker-synnada — APPROVED** [review]

**Comment by aturker-synnada** [comment]
LGTM!

**Pull request merged**